### PR TITLE
Enable pluggable components on Windows

### PR DESCRIPTION
**Pull request by philliphoff**
# Description

Removes the restriction on use of pluggable components on Windows.  Updates the Unix Domain Socket syntax and detection logic to handle Windows' quirks (namely using `unix:` instead of `unix://` and not relying on the `ModeSocket` flag).

In the case of Windows, as the `ModeSocket` flag is not set, the Dapr runtime instead relies on looking for "socket shaped" sockets which, by default, have the `.sock` extension. If the user desires to use another extension, they can set the `DAPR_COMPONENTS_SOCKETS_EXTENSION` environment variable to an arbitrary extension. (This environment variable is already supported in the Java and .NET pluggable component SDKs.)

With this change, the Dapr runtime will also respect temp directories specified via `$TMPDIR`, whereas it previously used `/tmp`.

## Issue reference

Please reference the issue this PR will close: #6082 (closed earlier for lack of comments)

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/components/pluggable/grpc.go [hunk 1]
```diff
@@ -103,6 +104,11 @@ func socketDialer(socket string, additionalOpts ...grpc.DialOption) GRPCConnecti
 // SocketDial creates a grpc connection using the given socket.
 func SocketDial(ctx context.Context, socket string, additionalOpts ...grpc.DialOption) (*grpc.ClientConn, error) {
 	udsSocket := "unix://" + socket
+
+	if runtime.GOOS == "windows" {
+		udsSocket = "unix:" + socket
```
Comment (daixiang0) [inline pkg/components/pluggable/grpc.go:109]: Could we add a test to cover it?

File: pkg/components/pluggable/discovery.go [hunk 2]
```diff
@@ -47,13 +47,16 @@ func removeExt(fileName string) string {
 }
 
 const (
-	SocketFolderEnvVar  = "DAPR_COMPONENTS_SOCKETS_FOLDER"
-	defaultSocketFolder = "/tmp/dapr-components-sockets"
+	SocketsFolderEnvVar  = "DAPR_COMPONENTS_SOCKETS_FOLDER"
+	defaultSocketsFolder = "dapr-components-sockets"
```
Comment (daixiang0) [inline pkg/components/pluggable/discovery.go:51]:
```suggestion
	DefaultSockets\Folder = "dapr-components-sockets"
```

File: utils/utils.go [hunk 3]
```diff
@@ -74,7 +81,15 @@ func GetIntValOrDefault(val int, defaultValue int) int {
 
 // IsSocket returns if the given file is a unix socket.
 func IsSocket(f fs.FileInfo) bool {
```
Comment (daixiang0) [inline utils/utils.go:83]: Please add tests for it.

**Review by daixiang0 — CHANGES_REQUESTED** [review]
@philliphoff Thank you for great contribution!

Just some nits here.

File: pkg/components/pluggable/discovery.go [hunk 4]
```diff
@@ -47,13 +47,16 @@ func removeExt(fileName string) string {
...
 )
 
-// GetSocketFolderPath returns the shared unix domain socket folder path
-func GetSocketFolderPath() string {
-	return utils.GetEnvOrElse(SocketFolderEnvVar, defaultSocketFolder)
+// GetSocketsFolderPath returns the shared unix domain socket folder path
+func GetSocketsFolderPath() string {
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply pkg/components/pluggable/discovery.go:55]: Happy to change it back, but the existing variable `DAPR_COMPONENTS_SOCKETS_FOLDER`, so thought the change would be more consistent (especially with `DAPR_COMPONENTS_SOCKETS_EXTENSION`).

File: pkg/components/pluggable/discovery.go (hunk 2 above)
Comment (philliphoff) [reply pkg/components/pluggable/discovery.go:51]: Wasn't sure if there was a reason for the existing naming, so didn't change it originally, but can do so.

File: pkg/components/pluggable/grpc.go (hunk 1 above)
Comment (philliphoff) [reply pkg/components/pluggable/grpc.go:109]: I'll see what can be added.

File: utils/utils.go (hunk 3 above)
Comment (philliphoff) [reply utils/utils.go:83]: PR has tests added to `SocketExists()` to cover the Windows cases, which *should* cover this function, being the only place it's called.

**Comment by JoshVanL** [comment]
@philliphoff please can you re-enable the socket integration test for windows in this PR? :) they are also quite limited so consider extending them a bit if you want
https://github.com/dapr/dapr/blob/53ceb4aa59a57a25de9d76b61b298584f5339354/tests/integration/suite/daprd/pluggable/basic.go#L48

**Comment by philliphoff** [comment]
> please can you re-enable the socket integration test for windows in this PR? :) they are also quite limited so consider extending them a bit if you want

Enabled those tests, which seem to pass on Windows and other platforms (though looks like there's some flakyness in other tests).

File: pkg/components/pluggable/grpc.go (hunk 1 above)
Comment (philliphoff) [reply pkg/components/pluggable/grpc.go:109]: Found some more related tests to unblock on Windows.

**Comment by philliphoff** [comment]
@daixiang0 Anything more you'd like me to add?

**Comment by m3nax** [comment]
Yesterday, version 0.3.0 of the dapr-sandbox/components-dotnet-sdk was released which includes the initialisation of Unix Domain Sockets on windows, can you enable the search in the component folder on windows as well so that people using Visual Studio can also develop on it?

**Comment by philliphoff** [comment]
> Yesterday, version 0.3.0 of the dapr-sandbox/components-dotnet-sdk was released which includes the correct initialisation of Unix Domain Sockets on windows, can you enable the search in the component folder on windows as well so that people using Visual Studio like me can also develop on it?

@m3nax Can you be a bit more specific about what you're trying to do?  I'm not sure what you mean by "search in the component folder" with respect to pluggable components.  (Usually the UDS is placed in a temp folder separate from the Dapr components folder.)

**Comment by m3nax** [comment]
@philliphoff It's quite simple, I'm using Visual Studio Dapr Extension for Visual Studio 2022 whitch starts debugging on my pc, whitch is Windows. The task manager confirm it.

image

 Then, when I try to register the binding to use it in another application I get the current message: 
``` 
level=debug msg="the current OS does not support pluggable components feature, skipping initialization" app_id=myapp
```

**Comment by philliphoff** [comment]
> This makes it impossible to debug pluggable components with Visual Studio 2022. Is there any other way?

@m3nax Until this PR gets committed, likely not.  the VS Dapr extension *does* support limited debugging of applications within WSL, but I've not tried to add a pluggable component into the mix so I can't say for certain whether that would work.

**Comment by m3nax** [comment]
Not stale

**Comment by philliphoff** [comment]
(Just a note that I'll be unlikely to have time in the near term to freshen up this PR, also given that there have been some significant changes for component hot-reload that intersect these changes.  I welcome any interested people to take that up if they have time/interest, else I'll probably just close this PR.)

**Pull request closed without merging**